Integration lifecycle model proposal

rhuss opened the pull request:
This WIP is published very early to gather some early feedback about the direction. The idea is to have a reconcilation model like for Kubernetes, but specific for integrations.

It doesn't yet contain ideas about versioning and how to include this into the state model.
Actual the most important parts are the static state model and the introduction of the two states used to reconcile an integration: currentState and targetState.

"Pending" actually is not a static state of the intergration but the state of the reconciliation process. However on the UI it might be intermixed with the other states or might be presented by some other means.

rhuss committed:
Add an very early draft on integration lifecycle proposal

rhuss committed:
Update formatting

rhuss committed:
Fixed name

rhuss committed:
Worked on, added versioning

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+
+Technical an Integration is the domain object describing a Camel Route. 
+It is managed by `syndesis-rest` and `syndesis-ui`. 
+During design time this object only lives in this domain. 
```
Question on this one, currently as an integration is created or changed in the UI those changes are only client-side *until* the user decides to actually save the integration.  However the original intent was that changes to the integration would actually be persisted to the backend in some fashion as they're made by the citizen user.

There's pros and cons to each approach, should we figure out which one to use as part of this proposal?

rhuss replied:
This section is a bit unclear. Actually I mean it lives only in the DB but there is not already an associated runtime (pod with camel routs) created.

I would distinguish between "Save" (store it in the DB, lightweight operation) and "Publish" (create the runtime deployment, heavyweight operation).

A "Publish" simply means then to change the `targetState` from "Draft" to "Active" in the DB stored integration. A background controller in syndesis-rest would then pick this up and create the pods. 

So imo the UI is still free to save either explicitely or implicitely for the 'save' operation. Only the "Publish" operation should be a dedicated step (== button).

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+Right after an integration is created it enters the _Draft_ state. This means 
+the integration is not yet deployed. An integration leaves the draft state either by being published or deleted (in which case the whole integration is removed)
+
+After an activation there is no way back to the _Draft_ state.
```
What does having a `draft` state give us?

Would this be an indicator of an 'unfinished' integration for example?

rhuss replied:
E.g. in _Draft_ the integration needs not to validate it simply a state for 'work-in-progress'. The most important motivation for not allowing it to go back to _Draft_ once out of the state is for a simplified operational mode as the integration then becomes immutable afterwards. When changing such an integration, a new Integration object with  the same name, but another version will be created, again in state _Draft_.

Yes, unfinished integrations are in state _Draft_.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+
+### Active
+
+An _active_ integration is a running integration. This state can be either created from a _Draft_ state when the integration gets published of from an _Inactive_ state when a suspended integration is started again.
```
nit:
> published of from

should be

> published from

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+
+### Undeployed
+
+An integration which was published is never deleted but can be undeployed. An undeployed integration can be re-deployed (and activated) if the user chose the version of this integration to be reactivated.
```
Would `undeployed` integrations show up in the UI anywhere?  How would the user actually totally delete an integration?

rhuss replied:
That's a very good question which we still need to discuss with @kcbabo . I could imagine to have some kind of "History" page for all integration with different versions. Since would contain all "undeployed" ones, at most one active integration and at most one integration in "Draft" (wip, this draft integration would become active when "Pubished", the active integration would get undeployed).

gashcrumb replied:
+1 we definitely could use an integration history page.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+
+### Error
+
+The error state indicates that the integration is not running. This state can be entered consciously but only as a side effect. An integration can enter the error state any time. 
```
Is it possible for an integration to be running but have a transient error?  Or let's say one connection of the integration winds up having a stale token, the integration is actually running, but is it processing messages (maybe) or do we plan on forcing the integration to just up and die on errors like this?

rhuss replied:
If possible I'd like to keep a single _Error_ state to keep it simple. This state can be left to _Active_ if the error is transient and can get fixed. Along with this state goes a `stateDescription` which can hold detailed information about the error, maybe with some hint how to fix it.

"Error" is not necessarily a final state.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+
+The following operations change the `targetState`. The `targetState` could be updated by simply changing the Integration's field, but it is recommended to use a more semantic rich REST API call for perfoming this actions (i.e. having specific endpoint which can be used by the UI to trigger the targetState change).
+
+* **Publish** sets the target state to _Active_. This operation can be called from the states:
```
What if you want to save but not update the running integration?  Or do we not support that?

rhuss replied:
If you save but not "publish" nothing will happen on the running integration.

When changing an integration the following steps and checks are performed. 
Please keep in mind that all integration objects with the same name belong together, they are just variants in different versions.

* If there is already an integration with this name in state _Draft_, update this integration.
* If there is no integration in _Draft_ but one in _Active_, the active integration is cloned to a new integration with the same name and in state _Draft_. This integration then can be changed freely and saved to the DB.
* When the changed integration gets published, it's `targetState` is set from "Draft" to "Active". 
* A controller in syndesis-rest detects, that integration with `currentState == Draft` and `targetStat == Active`. The controller then performs these operations in the background:
  - For an integration with `currentState == Active|Inactive` set `targetState = Undeployed`
  - Create / Update the GitHub repo
  - Trigger an OpenShift build  

(i added this to the document)

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+* Its cloned from the _active_ integration
+* The state of the clone is set to _Draft_
+* The `parentVersion` is set to the version of the active integration it is cloned from
+* The `version` is set to undefined.
```
This might answer my questions a bit on 'saving' :-)

gashcrumb replied:
Maybe actually this needs to be earlier in the document or referred to in the text leading up to this section?

rhuss replied:
Agreed, I'm going to rearrange the document a bit to clarify the difference between "Save" and "Publish".

gashcrumb replied:
Awesome, and tx for the clarification.  Definitely would like to migrate the UI to updating the integration as the user makes changes so this is all good stuff.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+    int version;
+  
+    // Message describing the state further (e.g. error message)
+    String stateMessage;
```
I feel like just a string for this may not be that great for communicating problems to the UI and the user if it's used for error messages too, WDYT?

rhuss replied:
Actually this class describes the domain model which is not necessarily that what arrives at the UI. I agree that we should strictly decouple user friendly message with technical error message. Any idea how we come from an integration error to a nice UI message ?

gashcrumb replied:
Gotcha, thanks for the clarification.  I guess currently if we're talking about the user looking at the integration list we'd be showing an error icon and most likely this state message will be reflected in the tooltip.  So that might do for a starting point, I mean, we're not expecting the citizen user to go and browse the logs for the pod, instead they'll be asking a more technical user to investigate.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+public class Integration {
+    
+    // Unique name of the integration
+    String name;
```
Should we maybe decouple the user-entered name from the git repo?  People will want to use names like "My awesome message processor".   Would we have a reference to the git repo URL at this level?

rhuss replied:
Good point, also when people want to rename an integration (we might then have to rename all integration instances). Maybe a surrogate, non-visible name would make sense here ?

Actually we could also model the Integration in two levels:

A single `Integration` which contains one or more `IntegrationInstance`s (with at most one active / inactive and most one in draft state).

The difference would be that this is a more explicit grouping instead of an implicit, label-like grouping via the name.

Any opinions what would be better ?

I just added an alternative model to the document.

rhuss replied:
Jus thinking a bit about this, I guess an explicit grouping with two objects makes more sense wrt to consistency. E.g. when changing a name, the repo or any other integration global property, then every such `Integration` object would need to be updated (if grouped via name or any other label like prop)

gashcrumb replied:
Yeah, `Integration` containing `IntegrationInstances` makes sense to me, and I definitely think a system-generated ID of some sort would be good.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+    IntegrationState currentState;
+    
+    // Target state of the integration    
+    IntegrationState targetState;
```
Using the same enum is less ideal, would it make sense to have a narrowed enum type for the target state?  As it doesn't make sense for tools to set this to `Error`.

rhuss replied:
An internal logic (like a monitoring component which uses custom health checks) could set this to `Error` if it want to mark the integration as invalid. 

For the interaction with the UI I think this 'currentState' and 'targetState' should not be exposed to the UI. Its the Rest endpoint which will set the states according to the required logic (e.g. the UI uses an endpoint "publish" and the Rest endpoint sets the targetState to "Active"). 

Of course we could also expose the reconciliation mechanism over the Rest endpoints, too (e.g. that the UI sets the targetState and simply saves the integration). In that case the endpoint could then validate for a valid target state (e.g. not to set the targetState to 'Error' from the UI.

tl;dr - I would like to keep both the same enum so that they are easier to compare and also the targetState could be in fact any state, too, when set internally.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+    
+    // The version of the integration which was the origin of this integration
+    // 0 if this is a new integration
+    int parentVersion;
```
Would commit IDs or tags be better for these version fields?

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+* A GitHub repository will be created if this is the first version of an integration (parentVersion == 0). 
+* If it is an new version of an existing integration (e.g. because a user changed something), a new Git branch is created from the old version's branch. This branch is named after the version.
+* The runtime files are created and commit to the Git repository
+* If this is an update of an existing integration, a currently running or suspended integration (`currentState == Active or Inactive`) is set to `targetState = Undeployed` (which triggers an undeployment on behalf of the controller)
```
This kinda seems then like you can't persist changes without publishing them right away, i.e. what if the user wants to make some updates, persist them but not make those changes active right away?

rhuss replied:
Publishing is a deliberate action. Before you publish you can save everything for the integration when it is in _Draft_. So the user can change everyhting without affecting the running integration. It's only after a "Publish" the machinery starts. 

Actually I'm not sure anymore about whether we should set the targetState to "Undeployed" actively. It would be better if the update of the currentState happens in this case when the OpenShift deployment undeploys the current integration in the course of deploying the new one.

I added the following question to the doc:

Question: It would be awesome if we could change the state for an _Active_ integration to _Undeployed_ by watching OpenShift for update deployments so that we can mark the old integration as _Undeployed_ and the new as _Active_. Is this possible with OpenShift ? How to correlate OpenShift deployments with integration versions ?

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,203 @@
+## UI changes and updates
+
+Within the UI the state of an integration should be visualized. 
+It is suggested that when the reconciliation state is `Ok` the _currentState_ should be shown and when the reconciliation state is `Pending` then "Pending" should be shown (possibly whith the `targetState` as a tool tip info). Alternatively, when a reconciliation process the label could be more specific about the action just performing ("Deploying", "Suspending", "Resuming", "Undeploying") which  ...
```
Yeah, if we maybe have the reconciliation state as a separate field on the integration that'd make the logic in the UI easier.  Most likely when it's `Pending` then we'll show the `currentState` using the appropriate visualization.  I guess though the reconciliation state attribute would have a third state, `Error` so we can show whatever error information is available.

rhuss committed:
Worked on

rhuss committed:
Revisited and emphasized the distinction between Integration and IntegrationRevision

rhuss commented:
@gashcrumb thanks a lot for all your comments !

I reworked the document, to make it hopefully a bit clearer. Also introduced `IntegrationRevision` which describes a specific version of an integration. An `Integration` is the bracket around all `IntegrationRevisions` and add some global, mutable information.

gashcrumb commented:
Of course!  Lemme re-read it now with your updates and see if I have any more to add...

rhuss committed:
Update images

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,271 @@
+
+| State       | Description |
+| ----------- |------------ |
+| Draft | Initial state of an integration. The `IntegrationRevision` is not yet deployed |
```
Maybe should be:

> Initial state of an `IntegrationRevision`?

rhuss replied:
gotcha

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,271 @@
+
+Both objects are managed by `syndesis-rest` and `syndesis-ui`. 
+
+When a user creates an integration, an overall `Integration` object is created as well as an `IntegrationRevision` which starts its life in the state _Draft_. 
```
I guess `IntegrationRevision` would always be created in the `Draft` state.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,271 @@
+| Inactive | `IntegrationRevision` is deployed but is not running|
+| Undeployed | An `IntegrationRevision` has been undeployed | 
+| Error | The `IntegrationRevision` is deployed but in an error state |
+
```
Tricky bit with the state here, `Integration` and `IntegrationRevision`, the `Integration`'s state is likely going to always be the most recent `IntegrationRevision`'s state _except_ when the most recent `IntegrationRevision` is in Draft.

rhuss replied:
Actually, the idea is to have no dedicated state for an Integration, but for individual IntegrationRevisions. The state of the overall Integration is then inferred from the revision's state (but for the UI this overall state is then already provided).

But you are right with respect to the `Error` state, it's a tricky one, because one should keep the original state which should be tried to reach again. But whenn the `state` and the `targetState` are the same, entering the `state` with `Error` will still keep the target state which then is tried to be reconciled.

gashcrumb commented on proposals/044-integration-model.md:
```diff
@@ -0,0 +1,271 @@
+| Draft | Initial state of an integration. The `IntegrationRevision` is not yet deployed |
+| Active | `IntegrationRevision` is deployed and running |
+| Inactive | `IntegrationRevision` is deployed but is not running|
+| Undeployed | An `IntegrationRevision` has been undeployed | 
```
Would this maybe be:

> All `IntegrationRevisions` have been undeployed

rhuss replied:
It's the individual IntegrationRevision which gets undeployed. Another revision can be still active.

The idea is:

* 0 or 1 revision in 'Active' or 'Inactive'
* N in 'Undeployed'
* 0 or 1 in Draft

rhuss committed:
Minor fixs

rhuss committed:
Update state diagram

rhuss committed:
doc: minor touches

rhuss commented:
Going to merge it now, so that we can easily start to work on it.
